### ISSUE TITLE
Which version of ImageMagick

### ISSUE THREAD

**Issue by vihari** [rkssisodiya/ascii_picture#1]
I tried installing the software on my machine, it is throwing some build errors with ImageMagick-7.0.1-10.
The version of ImageMagick is to be blamed I think, can you tell me the version of ImageMagick that you are using?

**Comment by rkssisodiya** [comment]
Hey @vihari its great to hear from you :) I am glad you found this one to be interesting. I should have tested this on ubuntu, now I have and have also modified the readme accordingly. Thanks again for the inputs
I tested on mac osx El Capitan 10.11.1 with imagemagick version 6.9.4 Q16 and on ubuntu 14.04 with imagemagick version 6.7.7 Q16 (on ubuntu the package needed for the headers is libmagick++-dev)

- The image does maintain the aspect ratio - its just that usually characters in our editors dont make for perfect squares (while pixels would) and hence it feels like the aspect ratio has changed when viewed in a text editor. Hope I understood your concern right.
- The image by default goes to 120 chars in width while maintaining the ratio because I felt that having more width than that by default leads to very large text output which then needs to be zoomed out a lot more than what it has to be at 120 chars. Even then, the width can be configured to retain the original size if we specify the width to be the original width in the command (though its just a hacky use). We can introduce an extra flag that gets this done and avoids resizing altogether, I will make a task out of it (You are more than welcome to send a PR with the change
- the smoothing idea sounds interesting but this was just a small hack that I put together so it seems too sophisticated to need that but maybe in the future.
- I didn't really understand the budget related suggestion, would be great if you could tell me more.

**Comment by vihari** [comment]
With ImageMagick-7, it throws the following error
> src/loader.cpp:45:11: error: no type named 'PixelPacket' in namespace 'Magick'; did you mean >'MagickCore::PixelPacket'?
>    const Magick::PixelPacket * pixels = image.getConstPixels(0, 0, width, height);

This issue is unsettling, I will try to patch it.

* I understand what you are trying to convey for aspect ratio problem, due to discrepancy in line and character spacing. I think the reason is something else. For example see the attached images below
![screen shot 2016-06-10 at 21 03 19](https://cloud.githubusercontent.com/assets/3615408/15969767/0a7ad8f0-2f4f-11e6-931e-8c8b0bfb822f.png)
![screen shot 2016-06-10 at 21 04 39](https://cloud.githubusercontent.com/assets/3615408/15969768/0adc04f4-2f4f-11e6-8951-6f2d8bfe2ece.png)<br>
The aspect ratio of jp2a generated ascii image seems fine to me. It looks as though ascii_picture tool is appending more lines than that are required. Can you please take a look and find if it is a bug in the code.
* The suggestion of width is not very important for now.
* My suggestion on smoothing and budget on number of chars are correlated. By budget, I mean an upper limit on number of chars that can be used to represent the image. For example, what if someone wants the ascii file to be small for some unearthly reason (so that it can be included in the code and not get the "small talk" from the manager :) ). Obviously, if the budget is less, then we will have to compromise some info. in the original image, in other words a large smoothing window. Also of concern to me is the cases where there is no limit on number of ascii chars that can be used, but required to conserve as much info. as possible. Consider the IIT Mandi-logo example, what if we want to see the text "Indian Institute of Tech.... " in the ascii image?

Hope I made clear my suggestions, I will try to make PLs for the same. 
Thanks

**Comment by rkssisodiya** [comment]
So I checked the source of jp2a, its awesome - there are a lot of refinements that have been done there. 2 things that pertain to the context here are
- jp2a has a much simpler and much better ascii character palette - replacing the palette in this code with the one used there gave much better results for the yoda image - just like you have posted above
- when going about resizing, jp2a does take into account, the fact that most of the characters in terminals and in general are twice as tall as they are wide and taking those factors into account leads to better  preservation of the aspect ratio - I will be looking into that when I get time.

Thanks for explaining your other points, I will do some reading into how smoothing is typically approached, I am not very well informed in the image processing area.

**Comment by vihari** [comment]
Yeah, the palette plays an important role in picture clarity. `Jp2a` also provides a command line option to configure the palette characters. 
It is very clever of Jp2a to consider for the fact about character height and width, great work catching it. Thanks for clearing it up.
